api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -215,22 +230,34 @@ message DeltaDiscoveryResponse {
 
   // Type URL for resources. Identifies the xDS API when muxing over ADS.
   // Must be consistent with the type_url in the Any within 'resources' if 'resources' is non-empty.
+  // This does not need to be set if *udpa_removed_resources* is used instead of
+  // *removed_resources*.
   string type_url = 4;
 
   // Resources names of resources that have be deleted and to be removed from the xDS Client.
   // Removed resources for missing resources can be ignored.
   repeated string removed_resources = 6;
 
+  // As with *removed_resources* but used when a removed resource was named in
+  // its *Resource*s with a *udpa.core.v1.ResourceName*.
+  // [#not-implemented-hide:]
+  repeated udpa.core.v1.ResourceName udpa_removed_resources = 7;
+
   // The nonce provides a way for DeltaDiscoveryRequests to uniquely
   // reference a DeltaDiscoveryResponse when (N)ACKing. The nonce is required.
   string nonce = 5;
 }
 
+// [#next-free-field: 6]
 message Resource {
   option (udpa.annotations.versioning).previous_message_type = "envoy.api.v2.Resource";
 
   // The resource's name, to distinguish it from others of the same type of resource.
-  string name = 3;
+  string name = 3 [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
+
+  // Used instead of *name* when a resource with a *udpa.core.v1.ResourceName* is delivered.
+  udpa.core.v1.ResourceName udpa_resource_name = 5
+      [(udpa.annotations.field_migrate).oneof_promotion = "name_specifier"];
 
   // The aliases are a list of other names that this resource can go by.
   repeated string aliases = 4;
```

Comment:
We're going to deprecate this, right?  Are we going to do that in a separate PR?

Reply:
Yes, separate PR. I've avoided adding any deprecations as the new alternatives are not ready yet, so wanted to avoid the "deprecated but not ready yet" anti-pattern ;-D.